orttraining/orttraining/test/optimizer/horizontal_parallel_test_utils.cc

```diff
@@ -125,9 +127,24 @@ void VerifyOutputs(const Tensor& expected_tensor, const Tensor& actual_tensor, b
                    float atol, float rtol, float threshold) {
   ASSERT_EQ(expected_tensor.Shape(), actual_tensor.Shape());
   auto size = expected_tensor.Shape().Size();
-  const std::vector<float> expected(expected_tensor.template Data<float>(), expected_tensor.template Data<float>() + size);
-  const std::vector<float> actual(actual_tensor.template Data<float>(), actual_tensor.template Data<float>() + size);
-  VerifyOutputs(expected, actual, use_threshold_compare, atol, rtol, threshold);
+  if (expected_tensor.GetElementType() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+    const std::vector<float> expected(expected_tensor.template Data<float>(), expected_tensor.template Data<float>() + size);
+    const std::vector<float> actual(actual_tensor.template Data<float>(), actual_tensor.template Data<float>() + size);
+    VerifyOutputs(expected, actual, use_threshold_compare, atol, rtol, threshold);
+  }
+#ifdef USE_CUDA
+  else if (expected_tensor.GetElementType() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) {
+    auto* expected = expected_tensor.template Data<MLFloat16>();
+    auto* actual = actual_tensor.template Data<MLFloat16>();
+  
+    std::vector<float> f_expected(size);
+    std::vector<float> f_actual(size);
+    ConvertMLFloat16ToFloat(expected, f_expected.data(), static_cast<int>(size));
+    ConvertMLFloat16ToFloat(actual, f_actual.data(), static_cast<int>(size));
+    VerifyOutputs(f_expected, f_actual, use_threshold_compare, math::halfToFloat(math::floatToHalf(atol)),
+                             math::halfToFloat(math::floatToHalf(rtol)), math::halfToFloat(math::floatToHalf(threshold)));
```

Comment:
>math::halfToFloat(math::floatToHalf [](start = 29, length = 35)

nit: indentation
why do we need to call halfToFloat(floatToHalf(x))?

Reply:
this is so that the tolerances/threshold go through the same loss of precision as the data

Reply:
why is it desirable to have the same loss of precision for tolerance/threshold?